src/data/program_configuration.js

```diff
@@ -129,31 +140,49 @@ class SourceExpressionBinder<T> implements Binder<T> {
         return [];
     }
 
-    populatePaintArray(length: number, feature: Feature) {
+    populatePaintArray(newLength: number, feature: Feature) {
         const paintArray = this.paintVertexArray;
 
         const start = paintArray.length;
-        paintArray.reserve(length);
+        paintArray.reserve(newLength);
 
         const value = this.expression.evaluate({zoom: 0}, feature);
 
         if (this.type === 'color') {
             const color = packColor(value);
-            for (let i = start; i < length; i++) {
+            for (let i = start; i < newLength; i++) {
                 paintArray.emplaceBack(color[0], color[1]);
             }
         } else {
-            for (let i = start; i < length; i++) {
+            for (let i = start; i < newLength; i++) {
                 paintArray.emplaceBack(value);
             }
 
             this.statistics.max = Math.max(this.statistics.max, value);
         }
     }
 
+    updatePaintArray(start: number, end: number, feature: Feature) {
+        const paintArray = this.paintVertexArray;
+        const value = this.expression.evaluate({zoom: 0}, feature);
+
+        if (this.type === 'color') {
+            const color = packColor(value);
+            for (let i = start; i < end; i++) {
+                paintArray.emplace(i, color[0], color[1]);
+            }
+        } else {
+            for (let i = start; i < end; i++) {
+                paintArray.emplace(i, value);
+            }
+
+            this.statistics.max = Math.max(this.statistics.max, value);
```

Comment:
Also, the max changing means the cached `queryPadding` value might be invalid: https://github.com/mapbox/mapbox-gl-js/blob/bd5a65380b1ffb961bd2e3173b0b96f8aa445f8f/src/source/tile.js#L170-L174

I think we need some way of either invalidating it, or if it is cheap enough, recalculating it every time we need to use it instead of caching it.

Reply:
Added this in 
https://github.com/mapbox/mapbox-gl-js/blob/25d23f946bbb65329fcbc8c5d07dc3ca8bb68717/src/source/tile.js#L435-L438

`Tile.queryPadding` is updated after each` bucket update.